pkg/controller/job/job_controller.go

```diff
@@ -393,7 +396,34 @@ func (jm *Controller) processNextWorkItem() bool {
 	}
 	defer jm.queue.Done(key)
 
+	keepInMetrics := true
+	mode := string(batch.NonIndexedCompletion)
+	ns, name, err := cache.SplitMetaNamespaceKey(key.(string))
+	if err != nil || len(ns) == 0 || len(name) == 0 {
+		keepInMetrics = false
+	} else {
+		job, err := jm.jobLister.Jobs(ns).Get(name)
+		if err != nil {
+			keepInMetrics = false
+		} else {
+			// make a copy so we don't mutate the shared cache
+			if isIndexedJob(job) {
+				mode = string(batch.IndexedCompletion)
+			}
+		}
+	}
+
+	now := time.Now()
 	forget, err := jm.syncHandler(key.(string))
+	if keepInMetrics {
+		metrics.JobSyncDurationSeconds.WithLabelValues(mode).Observe(time.Since(now).Seconds())
```

Comment:
I think we can do this inside `syncHandler` instead. That way, we don't have to query the joblister twice.
You can do it in a `defer` function.

Reply:
The `JobSyncNum` also needs the job `mode` information. If we don't want to query the joblister again outside the syncHandler, we need to add several lines to increase `JobSyncNum` in `syncHandler` where it returns an error or nil. 
I think it maybe more concise and clear to keep the current code, unless query the joblister costs a lot.

Reply:
JobSyncNum and JobSyncLatency should be updated in the same scenarios: when the job was not deleted.

Reply:
yep. and they both need to know if the syncJob returns an error or not. Use `defer` function can not get error information

Reply:
yes, it can :)

You have to name the return values